src/turbomind/kernels/apply_token_bitmask_inplace_cuda.cu

```diff
@@ -0,0 +1,253 @@
+namespace turbomind {
+using namespace turbomind::core;
+
+void ApplyTokenBitmaskInplace(Tensor logits, Tensor bitmask, std::optional<Tensor> indices)
+{
+    std::pair<int32_t, int32_t> logits_shape =
+        logits.ndim() == 2 ?
+            std::make_pair(static_cast<int32_t>(logits.shape(0)), static_cast<int32_t>(logits.shape(1))) :
+            std::make_pair(1, static_cast<int32_t>(logits.shape(0)));
+
+    std::pair<int32_t, int32_t> bitmask_shape =
+        bitmask.ndim() == 2 ?
+            std::make_pair(static_cast<int32_t>(bitmask.shape(0)), static_cast<int32_t>(bitmask.shape(1))) :
+            std::make_pair(1, static_cast<int32_t>(bitmask.shape(0)));
+
+    int vocab_size = std::min(logits_shape.second, bitmask_shape.second * BITS_PER_BLOCK);
+
+    int32_t  num_rows    = logits_shape.first;
+    int32_t* indices_ptr = nullptr;
+    if (indices) {
+        num_rows    = indices->shape(0);
+        indices_ptr = indices->data<int32_t>();
+    }
+    else {
+        TM_CHECK(logits_shape.first == bitmask_shape.first) << "logits and bitmask must have the same batch size.";
+    }
+
+    switch (logits.dtype()) {
+        case kFloat32: {
+            ApplyTokenBitmaskInplaceDispatchToPackedT(logits.data<float>(),
+                                                      bitmask.data<int32_t>(),
+                                                      indices_ptr,
+                                                      vocab_size,
+                                                      logits.stride(0),
+                                                      bitmask.stride(0),
+                                                      num_rows);
+            break;
+        }
+        case kFloat16: {
+            ApplyTokenBitmaskInplaceDispatchToPackedT(logits.data<half_t>(),
+                                                      bitmask.data<int32_t>(),
+                                                      indices_ptr,
+                                                      vocab_size,
+                                                      logits.stride(0),
+                                                      bitmask.stride(0),
+                                                      num_rows);
+            break;
+        }
+#if __CUDA_ARCH__ >= 800
+        case kBfloat16: {
+            ApplyTokenBitmaskInplaceDispatchToPackedT(logits.data<bfloat16_t>(),
+                                                      bitmask.data<int32_t>(),
+                                                      indices_ptr,
+                                                      vocab_size,
+                                                      logits.stride(0),
+                                                      bitmask.stride(0),
+                                                      num_rows);
+            break;
+        }
+#endif
+        default:
+            TM_CHECK(false) << "logits dtype must be float, float16 or bfloat16.";
+            break;
+    }
```

Comment:
May consider using `TM_DISPATCH_PRIMARY_DTYPES` for a more robust and maintainable dispatcher implementation.
Here is a reference: https://github.com/InternLM/lmdeploy/blob/d18ab5694cdf8b9ddac4bdee7bcf1b8794fc3114/src/turbomind/kernels/activation.cu#L100

Reply:
As we just use float logits, I have streamlined the code to just launch float kernel.